### PR TITLE
perf: separate generatingUUID logic & making it more complex

### LINKED ISSUE #8: Fix nested ChangeHighlight
As a user I might need to highlight a div and some times to highlight a small part of it, or maybe highlight another part differently, this can be done using nested `ChangeHighlight` like this:
```jsx
<ChangeHighlight>
    <div ref="ref1">
        {message1}
        <ChangeHighlight>
          <div ref="ref2">{message2}</div>
        </ChangeHighlight>
    </div>
</ChangeHighlight>
```
with the current implementation, the styling of the top `ChangeHighlight` will override the internal one's style which is not the desirable action in this case.

**Proposal:** we need during creating the shadowDOM to skip the ref element that are wrapped inside another `ChangeHighlight` element.

### PR DESCRIPTION
**Pull request by EslamHiko**
This is a PR to fix: https://github.com/medhatdawoud/react-change-highlight/issues/8

uuid

### CODE REVIEW

File: src/Util/generateUUID.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+const generateUUID = () => {
+  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
+    var r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
```
Comment (medhatdawoud) [inline src/Util/generateUUID.ts]: It seems to be less chances of having a duplicate id but it's ok, I'm fine with using bitwise despite I don't find a big benefit of using it here, just needs some cleaning to be kind of this:

```js
const randomNum = Math.random() * 16 | 0
const finalChar = c === 'x' ? randomNum : (randomNum & 0x3 | 0x8);
```

File: src/ChangeHighlight.tsx [hunk 2]
```diff
@@ -31,12 +33,12 @@ const ChangeHighlight: React.FC<Props> = ({
...
   const [shadowDOM, setShadowDOM] = useState([]);
   const isInitialMount = useRef(true);
-  const [uniqueId, setUniqueId] = useState(0);
+  const [uniqueId, setUniqueId] = useState("");
 
   useEffect(() => {
     if (!disabled) {
       if (isInitialMount.current) {
-        setUniqueId(Math.floor(Math.random() * 100000)+1);
+        setUniqueId(generateUUID());
```
Comment (medhatdawoud) [inline src/ChangeHighlight.tsx]: I prefer to change the function name to have the proper verb, so it's more convenient to have `get` in the beginning, something like this `getUUID()`

**Review by medhatdawoud — CHANGES_REQUESTED** [review]

File: src/Util/setHighlight.ts [hunk 3]
```diff
@@ -10,12 +10,12 @@ const setHighlight = (
...
   setClearHighlightFn: Function,
   updateClearHandler: Function,
-  uniqueId: number
+  uniqueId: string
 ) => {
 
   const element = child.ref.current;
 
-  if (!isInitial && element && parseInt(element.getAttribute(defaults.HIGHLIGHT_UNIQUE_ID)) === uniqueId) {
+  if (!isInitial && element && element.getAttribute(defaults.HIGHLIGHT_UNIQUE_ID) === uniqueId) {
```
Comment (medhatdawoud) [inline src/Util/setHighlight.ts]: I'v updated the version of typescript to `3.7.2` which has the optional chaining, so this like could be as follow
```js
if (!isInitial && element?.getAttribute(defaults.HIGHLIGHT_UNIQUE_ID) === uniqueId) {
```

**Review by medhatdawoud — APPROVED** [review]

**Comment by EslamHiko** [comment]
@medhatdawoud I would like to thank you so much for your content and I'll take your advice into action now, I'm looking for a job,  So can we have a small conversation about my skills & experience, and hopefully convince you to refer me to a good position?

**Comment by medhatdawoud** [comment]
@EslamHiko Thanks for your contribution, let's move that chat to be on twitter, send me a DM @med7atdawoud

**Pull request merged**